ddtrace/contrib/pymongo/patch.py

```diff
@@ -1,14 +1,49 @@
 import pymongo
 
+from ddtrace import Pin
+from ddtrace.vendor.wrapt import wrap_function_wrapper as _w
+
+from ...ext import mongo as mongox
+from ...utils.deprecation import deprecated
 from .client import TracedMongoClient
 
 # Original Client class
 _MongoClient = pymongo.MongoClient
 
-
 def patch():
+    # We should progressively get rid of TracedMongoClient. We now try to
+    # wrap methods individually. cf issue #_
     setattr(pymongo, 'MongoClient', TracedMongoClient)
-
+    patch_pymongo_module()
 
 def unpatch():
     setattr(pymongo, 'MongoClient', _MongoClient)
+    unpatch_pymongo_module()
+
+@deprecated(message='Use patching instead (see the docs).', version='1.0.0')
+def trace_mongo_client(client, tracer, service=mongox.SERVICE):
+    traced_client = TracedMongoClient(client)
+    patch_pymongo_module()
```

Comment:
I believe that we loose the possibility of patching individual clients with wrapper methods like `_w`, right?

Reply:
Right. The way we handle instance-specific configuration/state is to use the `Pin`. Instances can have their own corresponding `Pin` instances which the instrumentation will look for and use to determine how to instrument any given instance. We typically have a helper to provide this information to the instrumentation wrapper (see Flask for example: https://github.com/DataDog/dd-trace-py/blob/5e8141e594c0502a0e9b5c42b14885d0f4ae8346/ddtrace/contrib/flask/helpers.py#L13 and its usage: https://github.com/DataDog/dd-trace-py/blob/5e8141e594c0502a0e9b5c42b14885d0f4ae8346/ddtrace/contrib/flask/patch.py#L254-L260) Let me know if you have any further questions here.

This would be a breaking change, yes. We'd want to deprecate but keep support for the older method of instrumenting for manual usage, but use the new instrumentation by default if auto instrumenting.

Reply:
Oh I see below that you do this already :+1: